client/landing/stepper/index.tsx

```diff
@@ -0,0 +1,76 @@
+import '@automattic/calypso-polyfills';
+import accessibleFocus from '@automattic/accessible-focus';
+import { initializeAnalytics } from '@automattic/calypso-analytics';
+import config from '@automattic/calypso-config';
+import ReactDom from 'react-dom';
+import { BrowserRouter, Route, Switch, Redirect } from 'react-router-dom';
+import { LocaleContext } from '../gutenboarding/components/locale-context';
+import { WindowLocaleEffectManager } from '../gutenboarding/components/window-locale-effect-manager';
+import { setupWpDataDebug } from '../gutenboarding/devtools';
+import { useFlows } from './flows';
+
+import 'calypso/components/environment-badge/style.scss';
+
+function generateGetSuperProps() {
+	return () => ( {
+		environment: process.env.NODE_ENV,
+		environment_id: config( 'env_id' ),
+		site_id_label: 'wpcom',
+		client: config( 'client_slug' ),
+	} );
+}
+
+interface AppWindow extends Window {
+	BUILD_TARGET?: string;
+}
+
+declare const window: AppWindow;
+
+window.AppBoot = async () => {
+	setupWpDataDebug();
+	// User is left undefined here because the user account will not be created
+	// until after the user has completed the flow.
+	// This also saves us from having to pull in lib/user/user and it's dependencies.
+	initializeAnalytics( undefined, generateGetSuperProps() );
+	// Add accessible-focus listener.
+	accessibleFocus();
+
+	ReactDom.render(
+		<LocaleContext>
+			<WindowLocaleEffectManager />
+			<BrowserRouter basename="stepper">
+				<StepperFlows />
+			</BrowserRouter>
+		</LocaleContext>,
+		document.getElementById( 'wpcom' )
+	);
+};
+
+function StepperFlows() {
+	const NextHandler: React.FunctionComponent< { path: string } > = ( { path } ) => {
+		return <Redirect to={ path } />;
+	};
+
+	const flows = useFlows();
+
+	return (
+		<>
+			{ flows.map( ( flow ) => (
+				<Switch>
+					<Route exact path={ flow.path } key={ flow.path }>
+						<Redirect to={ [ ...flow.steps.values() ][ 0 ].path } />
+					</Route>
+					{ [ ...flow.steps.values() ].map( ( step ) => (
+						<Route exact path={ step.path } key={ `${ flow.path }-${ step.path }` }>
+							<flow.Render
+								Next={ NextHandler }
+								step={ step }
+								index={ flow.steps } // TS fail
```

Comment:
When I hover over `index=` it looks like it's expecting the type `FlowStepIndex<...> & FlowStepIndex<...>`. Notice the `&` operator. But when I hover over `flows.steps` it says it's a `FlowStepIndex<...> | FlowStepIndex<...>`. With the `|` operator.

I don't know exactly what's happening, but I think it's something to do with the inner render loops wanting to deal with one concrete flow type, but the type system still has all the information for all the possible flow types. Maybe because it can't guarantee that you're going to iterate through all the flows in the correct order?

I managed to move the type error to somewhere that I felt made the issue more obvious. I split the outer render loop into it's own function. It looks something like this.

```typescript
function StepperFlows() {
	const flows = useFlows();
	return <>{ flows.map( renderFlow ) }</>; // TS error is now here
}

function renderFlow< S extends string >( flow: Flow< S > ) {
	return (
		<Switch>
			<Route exact path={ flow.path } key={ flow.path }>
			...
```

This new approach of having both arbitrary steps _and_ arbitrary flows makes the typing a bit more complicated, and I'm not quite sure how to solve it. One approach to take is to say that `stepper/index.tsx` doesn't need all of that detailed type information so we can do some judicious type erasure. We could change `useFlows()` to

```typescript
export const useFlows = (): Flow< string >[] => {
	return [ RamboFlow, RamboLinearFlow ] as Flow< any >[];
};
```

The callers of `userFlows()` don't actually need to know any of the concrete flow types, it just needs to know that they are valid `Flow`s.

Reply:
Just found another way.

Add a `RenderProps` definition to `types.ts`

```typescript
export type RenderProps = Parameters< Flow< string >[ 'Render' ] >[ 0 ];
```

And then remove the explicit `Flow<...>` type from the flow definitions, using the `RenderProps` definition to tidy up the new TS errors.

```typescript
// 3. Flow config
const ramboLinearFlow = {
	path: '/rambo-linear',
	steps: ramboFlowIndex,
	Render: ( { Next, step, index }: RenderProps ) => {
		const [ next, setNext ] = useState< string | null >( null );
		...
```

Then do the same for `RamboFlow`.

This fixes the type error, probably by not overly-constraining the `ramboLinearFlow`. However, I'm worried that doing it this way will introduce more complicated type errors down the track. When devs are adding new flows, it can be useful to have the flow definition's type constrained so that you get auto complete. So part of me still likes the type erasure approach from above.

Reply:
Thanks @p-jackson . will take a look at these on Tuesday if none gets to. 

Just a note about this - if it helps - I doubt the "index" prop will remain being passed back into the flow/steps from here. Non-conclusive, but the idea brewing is active flows registering their indexes to a store and having access to them via selectors/hooks (useFlowIndex, useFlowNavigation{getNextStepInFlow}, etc).